### PR TITLE
Fixes that Blazor focus extension could not focus SVG elements

### LINKED ISSUE #35046: `Error: Unable to focus on an invalid element` when focusing on SVG elements
### Describe the bug
When you try to focus an SVG element that has a `tabindex` attribute an error is thrown from here: https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/src/DomWrapper.ts#L12

This is because SVG elements are not an instance of [HTMLElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement).

But SVG elements can actually be focused if they have a `tabindex` attribute like so:
```html
<svg height="50" width="50">
    <rect tabindex="0" x="5" y="5" width="40" height="40" fill="grey"></rect>
</svg>
```
This can be fixed by changing the function that checks to the following:
```typescript
function focus(element: HTMLElement, preventScroll: boolean): void {
  if (element instanceof HTMLElement) {
    element.focus({ preventScroll });
  } else if (element.hasAttribute('tabindex')) {
    element.focus({ preventScroll });
  } else {
    throw new Error('Unable to focus an invalid element.');
  }
}
```
This will check if the element has the `tabindex` attribute.

This might need more thought as that could be too loose a definition.

### To Reproduce
A repo that reproduces this error can be found here: https://github.com/KristofferStrube/BlazorSVGFocusBug/
The repo can be demoed on GitHub Pages here: https://kristofferstrube.github.io/BlazorSVGFocusBug/

### Exceptions (if any)
```bash
Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Unable to focus an invalid element.
      Error: Unable to focus an invalid element.
          at Object.focus (https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/blazor.webassembly.js:1:25215)
          at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/blazor.webassembly.js:1:2669
          at new Promise (<anonymous>)
          at Object.beginInvokeJSFromDotNet (https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/blazor.webassembly.js:1:2643)
          at Object.ut [as invokeJSFromDotNet] (https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/blazor.webassembly.js:1:52772)
          at _mono_wasm_invoke_js_blazor (https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.6.0.0-preview.6.21352.12.js:1:207806)
          at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.wasm:wasm-function[2461]:0x78fb6
          at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.wasm:wasm-function[2187]:0x6130b
          at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.wasm:wasm-function[6671]:0x15d556
          at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.wasm:wasm-function[6495]:0x154eb0
Microsoft.JSInterop.JSException: Unable to focus an invalid element.
Error: Unable to focus an invalid element.
    at Object.focus (https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/blazor.webassembly.js:1:25215)
    at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/blazor.webassembly.js:1:2669
    at new Promise (<anonymous>)
    at Object.beginInvokeJSFromDotNet (https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/blazor.webassembly.js:1:2643)
    at Object.ut [as invokeJSFromDotNet] (https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/blazor.webassembly.js:1:52772)
    at _mono_wasm_invoke_js_blazor (https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.6.0.0-preview.6.21352.12.js:1:207806)
    at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.wasm:wasm-function[2461]:0x78fb6
    at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.wasm:wasm-function[2187]:0x6130b
    at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.wasm:wasm-function[6671]:0x15d556
    at https://kristofferstrube.github.io/BlazorSVGFocusBug/_framework/dotnet.wasm:wasm-function[6495]:0x154eb0
   at Microsoft.JSInterop.JSRuntime.<InvokeAsync>d__16`1[[System.Object, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime , String , Object[] )
   at BlazorSVGFocusBug.Pages.Index.OnAfterRenderAsync(Boolean firstRender)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task , ComponentState )
```

#### ASP.NET Core version
6.0.100-preview.6.21355.2
#### Include the output of `dotnet --info`
```bash
.NET SDK (reflecting any global.json):
 Version:   6.0.100-preview.6.21355.2
 Commit:    7f8e0d76c0

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19041
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-preview.6.21355.2\

Host (useful for support):
  Version: 6.0.0-preview.6.21352.12
  Commit:  770d630b28

.NET SDKs installed:
  2.1.811 [C:\Program Files\dotnet\sdk]
  2.2.207 [C:\Program Files\dotnet\sdk]
  3.1.301 [C:\Program Files\dotnet\sdk]
  3.1.400-preview-015178 [C:\Program Files\dotnet\sdk]
  5.0.100 [C:\Program Files\dotnet\sdk]
  5.0.201 [C:\Program Files\dotnet\sdk]
  6.0.100-preview.6.21355.2 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.6.21355.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.5.21301.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.6.21352.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-preview.5.21301.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-preview.6.21353.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```
#### The IDE (VS / VS Code/ VS4Mac) you're running on, and its version
Visual Studio 2022 or dotnet CLI. It happens for both and when hosted on GitHub Pages.

### PR DESCRIPTION
**Pull request by KristofferStrube**
- [X] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [X] You've included unit or integration tests for your change, where applicable.
- [ ] You've included inline docs for your change, where applicable.
- [X] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

**PR Title**
Fixes that Blazor focus extension could not focus SVG elements

**PR Description**
I have added an extra case in the typescript wrapper for the focus extension that checks if the parsed element is an SVG element and also if it has a tabindex.
I have added a simple test that covers this behaviour.

Fixes #35046

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=35080) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=35080) before we can accept your contribution.<br/><br/>:x: KristofferStrube [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=35080)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=35080) it.</sub>

File: src/Components/test/testassets/BasicTestApp/SvgFocusComponent.razor [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿<svg width="100" heigth="100">
+    <circle tabindex="-1" cx="50" cy="50" r="20" id="focus-circle" @ref="circleReference" @onfocusin="@(() => { didReceiveFocusIn = true; })"></circle>
+</svg>
+
+<button id="focus-button" @onclick="@(() => FocusCircle(false))">Click to focus!</button>
+<hr />
+<p>Received focus in: <span id="focus-event-received">@didReceiveFocusIn</span></p>
+
+@code {
+    private ElementReference FocusCircle;
```
Comment (pranavkm) [inline src/Components/test/testassets/BasicTestApp/SvgFocusComponent.razor:10]:
```suggestion
    private ElementReference circleReference;
```

File: src/Components/Web.JS/src/DomWrapper.ts [hunk 2]
```diff
@@ -5,9 +5,15 @@ export const domFunctions = {
...
 
-function focus(element: HTMLElement, preventScroll: boolean): void {
+function focus(element: HTMLOrSVGElement, preventScroll: boolean): void {
   if (element instanceof HTMLElement) {
     element.focus({ preventScroll });
+  } else if (element instanceof SVGElement) {
+    if (element.hasAttribute('tabindex')) {
+      element.focus({ preventScroll });
+    } else {
+      throw new Error('Unable to focus an invalid SVG element because it does not have a tabindex.');
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/DomWrapper.ts:15]: Nit:
```suggestion
      throw new Error('Unable to focus an SVG element that does not have a tabindex.');
```
(The SVG element itself is not invalid, but it was just invalid to attempt to focus it.)

**Review by MackinnonBuck — COMMENTED** [review]
Looks great! 😃 Just a small wording suggestion.

**Review by MackinnonBuck — APPROVED** [review]

File: src/Components/test/testassets/BasicTestApp/SvgFocusComponent.razor (hunk 1 above)
Comment (MackinnonBuck) [inline src/Components/test/testassets/BasicTestApp/SvgFocusComponent.razor:10]:
```suggestion
    private ElementReference circleReference;
```

**Comment by KristofferStrube** [comment]
Many thanks for the help @MackinnonBuck 
I'm new to this so also had to learn when to pull upstream and when not to.

**Comment by KristofferStrube** [comment]
I was also a bit unsure if I had to commit the files in 
https://github.com/KristofferStrube/aspnetcore/tree/main/src/Components/Web.JS/dist/Release
that are generated when building.

**Comment by MackinnonBuck** [comment]
@KristofferStrube You've done a great job identifying the necessary changes 🙂 

> I was also a bit unsure if I had to commit the files in
https://github.com/KristofferStrube/aspnetcore/tree/main/src/Components/Web.JS/dist/Release
that are generated when building so I did not.

Yes, we usually do commit those files. CI might fail if you don't. Would you be able to do so?

**Comment by KristofferStrube** [comment]
> Yes, we usually do commit those files. CI might fail if you don't. Would you be able to do so?

Yep, I will get right on it but will have to build again locally first.

**Comment by KristofferStrube** [comment]
Hey @MackinnonBuck some of the tests failed due to what I think was a timeout or unlucky disc access errors. Can the CI tasks that failed be retriggered somehow or should I do something to fix this?

**Pull request merged**